web_monitoring/annotations_import.py

```diff
@@ -0,0 +1,116 @@
+    # Missing step: capture info from the "Who Found This" column and map it to
+    # the author's user record in the database eventually, but that requires a
+    # change to the API
+
+    bool_keys = [
+        'Language alteration',
+        'Content change/addition/removal',
+        'Link change/addition/removal',
+        'Repeated Change across many pages or a domain',
+        'Alteration within sections of a webpage',
+        'Alteration, removal, or addition of entire section(s) of a webpage',
+        'Alteration, removal, or addition of an entire webpage or document',
+        'Overhaul, removal, or addition of an entire website',
+        'Alteration, removal, or addition of datasets'
+    ]
+    string_keys = [
+        'Is this primarily a content or access change (or both)?',
+        'Brief Description',
+        'Topic 1',
+        'Subtopic 1a',
+        'Subtopic 1b',
+        'Topic 2',
+        'Subtopic 2a',
+        'Subtopic 2b',
+        'Topic 3',
+        'Subtopic 3a',
+        'Subtopic 3b',
+        'Any keywords to monitor (e.g. for term analyses)?',
+        'Further Notes',
+        'Ask/tell other working groups?'
+    ]
+    bools_dict = {k: csv_row[k] == '1' for k in bool_keys}
+    strings_dict = {k: csv_row[k] for k in string_keys}
+    annotation = {**bools_dict, **strings_dict}
+
+    significance = 0.0
+    if is_important_changes:
+        importance_significance_mapping = {
+            'low': 0.5,
+            'medium': 0.75,
+            'high': 1.0
+        }
+        row_importance = csv_row['Importance?'].lower().strip()
+        significance = importance_significance_mapping.get(row_importance, 0.0)
+    annotation['significance'] = significance
+
+    try:
+        annotation['priority'] = float(csv_row['Priority (algorithm)'])
+    except:
+        annotation['priority'] = 0.0
+
+    return annotation
+
+def post_annotation(change_ids, annotation):
+    cli = Client.from_env()
+    response = cli.add_annotation(annotation=annotation,
+                                  page_id=change_ids['page_id'],
+                                  to_version_id=change_ids['to_version_id'],
+                                  from_version_id=change_ids['from_version_id'])
+    return response
+
+def main():
+    doc = """Add analyst annotations from a csv file to the Web Monitoring db.
+
+Usage:
+path/to/annotations_import.py <csv_path> [--is_important_changes]
+
+Options:
+--is_important_changes  Was this CSV generated from an Important Changes sheet?
+"""
+    arguments = docopt(doc)
+    is_important_changes = arguments['--is_important_changes']
+    csv_path = arguments['<csv_path>']
+
+    # Missing step: Analyze CSV to determine spreadsheet schema version
+    for row in read_csv(csv_path):
+        change_ids = find_change_ids(row)
+        annotation = create_annotation(row, is_important_changes)
+        response = post_annotation(change_ids, annotation)
```

Comment:
I’m pretty sure this will raise a `db.WebMonitoringDbError` if there’s a problem with the POST. You might want to consider catching and logging it so the whole job doesn’t crash if there’s one bad row.

Reply:
Done, although from my simulated errors against a local server, it seems like db.py isn't properly catching/surfacing exceptions originating on the server

Reply:
If you want to verify this and make an issue for it, I can look into that as well

Reply:
I just tried this and it seems to:

```py
>>> from web_monitoring.db import Client
>>> Client.from_env().add_annotation(annotation={}, page_id='c4328d30-cada-452f-8642-4bff721f5fc2', from_version_id='a8a346fd-e275-498f-9463-11e1b88beb71', to_version_id='77ceecf5-a5a0-4ea6-9a05-afe79f85b356')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/rbrackett/Dev/datarescue/processing/web_monitoring/db.py", line 679, in add_annotation
    _process_errors(res)
  File "/Users/rbrackett/Dev/datarescue/processing/web_monitoring/db.py", line 38, in _process_errors
    raise WebMonitoringDbError(', '.join(map(repr, errors)))
web_monitoring.db.WebMonitoringDbError: {'status': 422, 'title': 'Annotations cannot be empty'}
```

(…although we really should be parsing more useful info out of that…)

Did you have a case where it should have but didn’t?

Reply:
It seems like 404s, e.g. annotations POSTs with page IDs which aren't in the server's DB, get raised as HTTPError (and there's a json decode error as well, for a blank response I presume):

```python
Traceback (most recent call last):
  File "/home/roto/edgi-govdata-archiving/web-monitoring-processing/web_monitoring/db.py", line 34, in _process_errors
    errors = res.json()['errors']
  File "/usr/lib/python3.7/site-packages/requests/models.py", line 897, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "scripts/annotations_import.py", line 130, in <module>
    main()
  File "scripts/annotations_import.py", line 123, in main
    response = post_annotation(change_ids, annotation)
  File "scripts/annotations_import.py", line 97, in post_annotation
    from_version_id=change_ids['from_version_id'])
  File "/home/roto/edgi-govdata-archiving/web-monitoring-processing/web_monitoring/db.py", line 679, in add_annotation
    _process_errors(res)
  File "/home/roto/edgi-govdata-archiving/web-monitoring-processing/web_monitoring/db.py", line 36, in _process_errors
    res.raise_for_status()
  File "/usr/lib/python3.7/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://localhost:3000/api/v0/pages/ca5f40d4-89bf-4d7d-bf1f-e89470f0c1b4/changes/..1fdfcee3-62ea-4bdb-9130-c3b289704dfc/annotations
```

Reply:
Oh, that is because you are hitting a server in dev mode instead of production mode. In dev mode, you get a fancy error page instead of JSON: https://github.com/edgi-govdata-archiving/web-monitoring-db/blob/e6a98a3f6e0512962d60075d36ce75ee4e728433/app/controllers/api/v0/api_controller.rb#L6

Try a bad production or staging URL, like: https://api-staging.monitoring.envirodatagov.org/api/v0/pages/ca5f40d4-89bf-4d7d-bf1f-e89470f0c1b4/changes/..1fdfcee3-62ea-4bdb-9130-c3b289704dfcd/annotations

On the other hand, it would be nice if:
- The DB server responded with a JSON error if you use the `Accept: application/json` header, regardless of dev vs. production mode.
- `db.py` in this repo actually set that header (which really is an oversight; it should always have been doing that).
- Debatable: should we really fall back to `HTTPError` if the response fails to parse, or still use `WebMonitoringDbError` in that case, just with a special message? It would certainly be more stable if we ever wanted to change HTTP libraries under the hood. (Any thoughts on this?)

But all of those are probably out of scope for this PR.

Reply:
Wrote two issues around that: edgi-govdata-archiving/web-monitoring-db#614 and #495.